src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,358 @@
+            _namespaces = new Dictionary<DPath, List<T>>();
+            _enums = new List<string>();
+            _count = 0;
+        }
+
+        internal TexlFunctionSet(T function)
+            : this()
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException($"{nameof(function)} cannot be null", nameof(function));
+            }
+
+            _functions.Add(function.Name, new List<T>() { function });
+            _functionsInvariant.Add(function.LocaleInvariantName, new List<T>() { function });
+            _namespaces.Add(function.Namespace, new List<T> { function });
+            _enums = function.GetRequiredEnumNames().ToList();
+            _count = 1;
+        }
+
+        internal TexlFunctionSet(IEnumerable<T> functions)
+            : this()
+        {
+            if (functions == null)
+            {
+                throw new ArgumentNullException($"{nameof(functions)} cannot be null", nameof(functions));
+            }
+
+            foreach (var func in functions)
+            {
+                Add(func);
+            }
+        }
+
+        private TexlFunctionSet(Dictionary<string, List<T>> functions, Dictionary<string, List<T>> functionsInvariant, Dictionary<DPath, List<T>> functionNamespaces, List<string> enums, int count)
+        {
+            _functions = functions ?? throw new ArgumentNullException($"{nameof(functions)} cannot be null", nameof(functions));
+            _functionsInvariant = functionsInvariant ?? throw new ArgumentNullException($"{nameof(functionsInvariant)} cannot be null", nameof(functionsInvariant));
+            _namespaces = functionNamespaces ?? throw new ArgumentNullException($"{nameof(functionNamespaces)} cannot be null", nameof(functionNamespaces));
+            _enums = enums ?? throw new ArgumentNullException($"{nameof(enums)} cannot be null", nameof(enums));
+            _count = count;
+        }
+
+        internal TexlFunctionSet(TexlFunctionSet<T> other)
+        {
+            _functions = new Dictionary<string, List<T>>(other._functions);
+            _functionsInvariant = new Dictionary<string, List<T>>(other._functionsInvariant);
+            _namespaces = new Dictionary<DPath, List<T>>(other._namespaces);
+            _enums = new List<string>(other._enums);
+            _count = other._count;
+        }
+
+        internal TexlFunctionSet(IEnumerable<TexlFunctionSet<T>> functionSets)
+            : this()
+        {
+            if (functionSets == null)
+            {
+                throw new ArgumentNullException($"{nameof(functionSets)} cannot be null", nameof(functionSets));
+            }
+
+            foreach (var functionSet in functionSets)
+            {
+                Add(functionSet);
+            }
+        }
+
+        // Slow API, only use for backward compatibility
+        [Obsolete("Slow API. Prefer using With* functions for idenfitying functions you need.")]
```

Comment:
>idenfitying [](http://example.com/codeflow?start=62&length=11)

nit - typo. 

+1 to making this [Obsolete]

Reply:
note that we still need some way to "list all functions", such as for REPL help or IDE.

Reply:
I'll add a FunctionNames getter and we'll see if we need more